Enable many complex number tests

MichaelTiemannOSC opened the pull request:
These changes put complex128 on an even footing with float64 and int64 as far as numerical testing is concerned.  These changes have been tested both against the Pandas test suite as well as the Pint-Pandas testsuite (using complex magnitudes).

These changes are a simpler version of a previous pull-request that was destroyed by GitHub's fork synchronize behavior.

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

MichaelTiemannOSC committed:
Enable many complex number tests

These changes put complex128 on an even footing with float64 and int64 as far as numerical testing is concerned.  These changes have been tested both against the Pandas test suite as well as the Pint-Pandas testsuite (using complex magnitudes).

These changes are a simpler version of a previous pull-request that was destroyed by GitHub's fork synchronize behavior.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update v2.1.0.rst

Add description of this PR to what's new file.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
While these simplified changes work great for complex128 (which is their job) they are not sufficient to also help with Pint-Pandas uncertainties, which is part of the motivation.  Reverting to draft while I investigate why not.

MichaelTiemannOSC commented:
Tests with uncertainties all pass (with similar `request`-related changes).  I cannot commit those changes to Pint-Pandas until these changes are accepted.  It is very nice that the underlying changes made to Pandas 2.1 really simplifies the Pint-Pandas uses of EAs.

@andrewgsavage @hgrecco

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Fix merge error in test_decimal.py

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Simplify test_fillna_no_op_returns_copy

xfail complex tests, but otherwise defer to parent object to implement test case.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
@mroeschke can you have a look at this and let me know if any additional changes are needed?

MichaelTiemannOSC committed:
Merge remote-tracking branch 'upstream/main' into test_numpy_complex2

mroeschke commented on pandas/core/dtypes/astype.py:
```diff
@@ -100,6 +100,11 @@ def _astype_nansafe(
 
+    elif np.issubdtype(arr.dtype, np.complexfloating) and is_object_dtype(dtype):
+        if np.isnan(arr).any():
+            res = np.asarray([np.nan if np.isnan(x) else x for x in arr], dtype)
```
Could you just do `res[np.isnan(arr)] = np.nan`?

MichaelTiemannOSC replied:
`res` has to be set to something first (it's the "result").  But I can look at using `copy` to decide if we really need something fresh before we do something that may have a side-effect.

mroeschke commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -864,6 +864,7 @@ Metadata
 
 Other
 ^^^^^
+- Add complex128 to the types of numerical data we test across the test suite (:issue:`54761`)
```
Is there an actual bug that was fixed? It would be better to describe that.

Also could you move to 2.2.0.rst if there is one

MichaelTiemannOSC replied:
Alas the number 54761 is the pull request itself, not an actual bug.  All this PR does it to add the complex128 datatype to the numerical data types in the test suite (and xfailing in the rare cases where there are obvious holes, such as the lack of a cython implementation for backfilling complex numbers if libs/algos.pxd).  So technically this is an ENH, except the only functionality it is adding is to the test suite.  All the complex128 machinery is already in Pandas.

MichaelTiemannOSC replied:
I think we decided that because there is no user-visible functionality change, that there's nothing to report in whatsnew, and I've taken that out.  So this is resolved.

mroeschke commented on pandas/tests/extension/base/missing.py:
```diff
@@ -86,7 +86,7 @@ def test_fillna_limit_backfill(self, data_missing):
         tm.assert_series_equal(result, expected)
 
-    def test_fillna_no_op_returns_copy(self, data):
+    def test_fillna_no_op_returns_copy(self, data, request):
```
These `request` don't seem used? (and below)

MichaelTiemannOSC replied:
It's not needed in the base class, but test_numpy.py and test_sparse.py need to reference request, and in order to do so, they need the base class to declare it as a parameter.

mroeschke replied:
> and in order to do so, they need the base class to declare it as a parameter

Hmm really? I thought pytest was not picky about request being referenced in a subclass method?

MichaelTiemannOSC replied:
Ah...lemme call the superclass without request argument (which it doesn't need).

MichaelTiemannOSC replied:
That worked, and also cleaned up decimal/test_decimal as well.

MichaelTiemannOSC committed:
changes from review

Attempt to resolve all comments from @mroeschke

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge remote-tracking branch 'upstream/main' into test_numpy_complex2

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

mroeschke commented on doc/source/whatsnew/v2.2.0.rst:
```diff
@@ -351,6 +351,7 @@ Styler
 
 Other
 ^^^^^
+- Add complex128 to the types of numerical data we test across the test suite (:issue:`54761`)
```
```suggestion
```

Not needed, unless we can describe a user facing behavior being changed

MichaelTiemannOSC replied:
OK.  I've deleted in my source and will commit with other potential changes in response to your comments.  No need to burn excess CI/CD cycles.

mroeschke commented on pandas/tests/extension/base/ops.py:
```diff
@@ -19,7 +19,7 @@ class BaseOpsUtil:
 
     def _get_expected_exception(
-        self, op_name: str, obj, other
+        self, op_name: str, obj, other, request
```
The `request` in this file don't look like they are used?

MichaelTiemannOSC replied:
Fixed by using LSP parameters elsewhere.

mroeschke commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -67,7 +67,7 @@ def data_for_grouping():
 class TestDecimalArray(base.ExtensionTests):
     def _get_expected_exception(
-        self, op_name: str, obj, other
+        self, op_name: str, obj, other, request
```
The `request` in this file don't look like they are used?

MichaelTiemannOSC replied:
Fixed by using LSP parameters elsewhere.

mroeschke commented on pandas/tests/extension/test_categorical.py:
```diff
@@ -153,7 +153,7 @@ def test_arith_frame_with_scalar(self, data, all_arithmetic_operators, request):
                 )
             )
-        super().test_arith_frame_with_scalar(data, op_name)
+        super().test_arith_frame_with_scalar(data, op_name, request)
```
The request in this file don't look like they are used?

mroeschke replied:
(and same generally throughout this PR)?

MichaelTiemannOSC replied:
I've just pushed some changes to use LSP (Linkov Substitution Principle)-style, which may strike you as more Pythonic.  It definitely cleans up the cases that offended, at a small cost of using `*args, **kwargs` instead of `request`.

MichaelTiemannOSC committed:
Use LSP parameter style for request

Replace `request` parameter with `*args, **kwargs` in many places.  This allows us to avoid needlessly passing request parameters just to satisfy method signatures.

Also remove whatsnew entry as this enhancement to test cases is not really user-visible.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
I strongly suspect the mypy failure was due to unrelated issues in CI/CD.  In the intervening weeks, things have moved forward.  Is it preferred practice to update via merge or rebase to catch up to latest after sitting idle for 3 weeks?

mroeschke commented:
Merging in main is sufficient. History is squashed during merging into main

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Handle complex128 EA in _ensure_data

The _duplicate functions expect complex128 data in ndarrays, not EAs.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Fix mypy pre-commit problems

The command `pre-comment run -a` doesn't do all the things that Pandas CI/CD does.  This commit fixes two problems found in the mypy section:

* Adding some comments to ignore some errors in _ensure_data for complex dtypes
* Fix the type signature of _get_expected_exception to match LSP stylings

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
Fixed mypy issues I did not previously understand, and caught a new bug introduced by #55255.  Accepting these changes might catch more bugs related to complex128 in the future ;-)

mroeschke commented on pandas/tests/extension/test_period.py:
```diff
@@ -72,10 +72,10 @@ def data_for_grouping(dtype):
 
 class TestPeriodArray(base.ExtensionTests):
-    def _get_expected_exception(self, op_name, obj, other):
+    def _get_expected_exception(self, op_name, obj, other, *args, **kwargs):
```
Why is args, kwargs needed here?

MichaelTiemannOSC commented on pandas/core/algorithms.py:
```diff
@@ -167,6 +167,11 @@ def _ensure_data(values: ArrayLike) -> np.ndarray:
+        # has no attribute "itemsize"
+        if values.dtype.itemsize == 16:  # type: ignore[union-attr]
+            # We have support for complex128
+            return np.asarray(values)
```
Interesting.  The cause of this comes from the initialization of `data` for the complex128 case:

```
 88  	@pytest.fixture
 89  	def data(allow_in_pandas, dtype):
 90  	    if dtype.numpy_dtype == "object":
 91  	        return pd.Series([(i,) for i in range(100)]).array
 92  	    breakpoint()
 93  ->	    arr = np.arange(1, 101, dtype=dtype._dtype)
 94  	    if dtype.kind == "c":
 95  	        arr = arr + (arr * (0 + 1j))
 96  	    return NumpyExtensionArray(arr)
```

The complex128 case is just doing what the other datatypes would be doing, and it happily generates such an EA:

```
(Pdb) NumpyExtensionArray(arr)
<NumpyExtensionArray>
[    (1+1j),     (2+2j),     (3+3j),     (4+4j),     (5+5j),     (6+6j),
     (7+7j),     (8+8j),     (9+9j),   (10+10j),   (11+11j),   (12+12j),
   (13+13j),   (14+14j),   (15+15j),   (16+16j),   (17+17j),   (18+18j),
   (19+19j),   (20+20j),   (21+21j),   (22+22j),   (23+23j),   (24+24j),
   (25+25j),   (26+26j),   (27+27j),   (28+28j),   (29+29j),   (30+30j),
   (31+31j),   (32+32j),   (33+33j),   (34+34j),   (35+35j),   (36+36j),
   (37+37j),   (38+38j),   (39+39j),   (40+40j),   (41+41j),   (42+42j),
   (43+43j),   (44+44j),   (45+45j),   (46+46j),   (47+47j),   (48+48j),
   (49+49j),   (50+50j),   (51+51j),   (52+52j),   (53+53j),   (54+54j),
   (55+55j),   (56+56j),   (57+57j),   (58+58j),   (59+59j),   (60+60j),
   (61+61j),   (62+62j),   (63+63j),   (64+64j),   (65+65j),   (66+66j),
   (67+67j),   (68+68j),   (69+69j),   (70+70j),   (71+71j),   (72+72j),
   (73+73j),   (74+74j),   (75+75j),   (76+76j),   (77+77j),   (78+78j),
   (79+79j),   (80+80j),   (81+81j),   (82+82j),   (83+83j),   (84+84j),
   (85+85j),   (86+86j),   (87+87j),   (88+88j),   (89+89j),   (90+90j),
   (91+91j),   (92+92j),   (93+93j),   (94+94j),   (95+95j),   (96+96j),
   (97+97j),   (98+98j),   (99+99j), (100+100j)]
Length: 100, dtype: complex128
```

mroeschke replied:
Ah okay thanks. Sure okay to keep this `np.asarray` call then

MichaelTiemannOSC committed:
Remove some LSP sigs for _get_expected_exception

Test functions defined in `BaseArithmeticOpsTests` such as `test_arith_series_with_scalar` call `check_opname` which calls `_get_expected_exception`.  Because `TestArithmetic` in `test_numpy.py` overrides `_get_expected_exception` and needs to access the `request` parameter, we need a mechanism to pass `request` down and back up the type hiearchy, which we can do using LSP style parameters `*args` and `**kwargs`.

We could opt to do this consistently across all EA types, but since only `decimal/test_decimal.py` overrides, and because it doesn't actually use the `result` parameter, we instead simply omit `*args, **kwargs` when not needed.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Additional `requests` removed; indentation fix

In `test_pyarrow.py`, fix indentation error in `TestArrowArray.test_arith_frame_with_scalar`

In `test_sparse.py`, remove unneeded `request` parameters in `TestArithmeticOps.test_arith_series_with_scalar` and `TestArithmeticOps.test_arith_series_with_array`

In `test_string.py`, remove unneeded `request` parameter in `TestMissing.test_fillna_no_op_returns_copy` and `Test2DCompat.arrow_not_supported`

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
I am mystified by the failure in `Unit Tests / ubuntu-latest actions-39.yaml`  I have attempted to recreate on my own machine using act, Docker, and my own repo and the `test_unique_bad_unicode` test cases pass.  They fail only in this one case, succeeding in ubuntu 3.10, 3.11, macos everywhere, windows everywhere, python-dev, etc.  Any tips on how to diagnose are welcome.  Also mystifying: nobody else is hitting this problem in their Unit tests.  Nor does this test go through any of the code I have modified (that I can see).  A Heisenbug?

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC commented:
One possible reason I cannot reproduce is because Pandas CI/CD Docker is running Linux 6.2.0 on Azure and the Docker I am running uses Linux 6.4.16-linuxkit on OSX.  I have tried other versions of Docker Desktop that have kernels like `5.15.49-linuxkit-pr` and `6.3.13-linuxkit` but neither of these reproduce the problem.

MichaelTiemannOSC committed:
Keep rval refs alive in StringHashTable._unique

Address a Heisenbug caused by `v = get_c_string(<str>repr(val))` potentially pointed to a string that is unreferenced the next time an exception is raised.  (Two exceptions are raised in succession in `pandas/tests/base/test_unique.py test_unique_bad_unicode`.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

WillAyd commented on pandas/_libs/hashtable_class_helper.pxi.in:
```diff
@@ -1128,6 +1128,7 @@ cdef class StringHashTable(HashTable):
         use_na_value = na_value is not None
 
         # assign pointers and pre-filter out missing (if ignore_na)
+        keep_rval_refs = []
```
I see this is also in #55530 - can you remove from this PR so not accidentally merged?

MichaelTiemannOSC replied:
I'll convert that to draft and then await approval of this PR and merge these changes into 55530 and re-open.  If I merely remove this fix from 55530 it will fail and waste cycles in the process.

MichaelTiemannOSC replied:
Given all the Cython 3.0 changes, I'll revert this change and see what happens with the heisenbug.

jbrockmendel commented on pandas/tests/arithmetic/test_numeric.py:
```diff
@@ -1082,45 +1082,75 @@ def test_ufunc_compat(self, holder, dtype):
         result = np.sqrt(idx)
-        assert result.dtype == "f8" and isinstance(result, box)
-        exp = Index(np.sqrt(np.array([1, 2, 3, 4, 5], dtype=np.float64)), name="x")
+        if result.dtype.kind == "c":
```
> that lets us cast to the nearest inexact data type

I don't think so, no.  i expected maybe_promote to do that, but looks like it always gives float64

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC commented on pandas/tests/extension/test_sparse.py:
```diff
@@ -371,13 +381,13 @@ def _skip_if_different_combine(self, data):
             pytest.skip("Incorrected expected from Series.combine and tested elsewhere")
 
-    def test_arith_series_with_scalar(self, data, all_arithmetic_operators):
+    def test_arith_series_with_scalar(self, data, all_arithmetic_operators, request):
```
Fixed by using LSP parameters elsewhere.

MichaelTiemannOSC commented on pandas/core/nanops.py:
```diff
@@ -1003,11 +1003,25 @@ def nanvar(
-    # See https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
-    avg = _ensure_numeric(values.sum(axis=axis, dtype=np.float64)) / count
+    # See https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance...
+    if values.dtype.kind == "c":
```
`values` could be boolean, in which case `values.sum` computes one thing for dtype=values.dtype and another for values.float64 (the latter presumed correct).  But when `values` are complex we need to compute `avg` based on the complex dtype, not `np.float64`.  I'll refactor the code to test values.dtype.kind only once.

MichaelTiemannOSC commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -337,7 +337,8 @@ def test_setitem_slice_array(self, data):
     def test_setitem_scalar_key_sequence_raise(self, data):
         arr = data[:5].copy()
-        with pytest.raises(ValueError):
+        # complex128 data raises TypeError; other numeric types raise ValueError
```
When running `pandas/tests/extension/test_numpy.py` and `data` is an array of `float64` we see:
```
ValueError: setting an array element with a sequence.
```
When `data` is `complex128` we see:
```
TypeError: must be real number, not NumpyExtensionArray
```

The ValueError seems to come from Numpy, and the TypeError seems to come from Python, both coming from `__setitem__` in class `NDArrayBackedExtensionArray` in `pandas/core/arrays/_mixins.py`.

mroeschke replied:
Could we catch the Python TypeError and reraise as a ValueError?

MichaelTiemannOSC replied:
I've amended the PR to attempt this.  Please let me know what you think.

MichaelTiemannOSC committed:
Code review changes

Changes suggested by @WillAyd

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Fix incomplete removal of `keep_rval_refs`

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge remote-tracking branch 'upstream/main' into test_numpy_complex2

Trying to keep these changes fresh enough to review and merge.  The should be good to go.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Merge remote-tracking branch 'upstream/main' into test_numpy_complex2

MichaelTiemannOSC committed:
Update io.py

Clean up random merge conflict.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update test_numpy.py

Instead of overriding _get_expected_exception, set expected exception types in various test cases.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update test_numpy.py

Fix indentation errors.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Update ops.py

Don't need `*args, **kwargs` changes anymore due to refactoring upstream.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update test_decimal.py

Re-harmonize with upstream.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Further simplifications due to upstream

It turns out that `data` in `test_sparse.py` never contains complex numbers, and furthermore, that it's quite a lot of extra work to make the sparse tests complex-friendly.  So we leave complex number testing out of test_sparse.  Contributions welcome, as usual.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update test_arrow.py

Arrow doesn't support complex numbers, so no need to special case tests as if it does.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update test_arrow.py

Restore function accidentally deleted.  Intention was to only delete unneeded complex code handling, not the whole function.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Update test_arrow.py

Delete `request` parameter no longer needed for `test_fillna_no_op_returns_copy`.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
Thanks to some upstream changes, the changes required to support many complex test cases is greatly simplified.  Please let me know your feedback on these changes.

MichaelTiemannOSC committed:
setitem exceptions for complex raise ValueError

In certain cases where Python throws a TypeError for complex values where it would throw ValueError for real values, transform exception to ValueError.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC commented:
Can somebody remove the stale tag from this PR?

MichaelTiemannOSC commented:
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

I have updated this PR.

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Update _mixins.py

Make ruff happy by raising from `exc`.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
Note that the only failure now is Numpy Dev (and most likely "not my fault").

jbrockmendel commented on pandas/core/arrays/_mixins.py:
```diff
@@ -259,7 +259,15 @@ def shift(self, periods: int = 1, fill_value=None):
+            # Don't let Python's handling of `complex` make extra complexity for Pandas
+            if self._ndarray.dtype.kind == "c":
+                raise ValueError(
+                    *(x.replace("real", "complex") for x in exc.args)
```
why are we tinkering with the exception message?

MichaelTiemannOSC replied:
Maybe this is actually a NumPy bug.  When `value` is an object, NumPy appears to try constructing a complex number by passing the object in as the real component, but that, too fails with the message `TypeError: must be a real number, not object`.  But that's a wrong message, because what we really want is to report the failure to construct a complex number.  Here's a test case of sorts:
```
import numpy as np
xx = object()
np.asarray(xx).astype('float64')
# *** TypeError: float() argument must be a string or a real number, not 'object'
np.asarray(xx).astype('complex128')
# *** TypeError: must be real number, not object
```
That's what I'm trying to fix.  So I could file an issue against NumPy, remove the above special case, and Pandas will fix itself when NumPy fixes itself.  Is that the best approach?

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -170,6 +170,11 @@ def _ensure_data(values: ArrayLike) -> np.ndarray:
+        # error: Item "ExtensionDtype" of "Union[Any, ExtensionDtype]"
+        # has no attribute "itemsize"
+        if values.dtype.itemsize == 16:  # type: ignore[union-attr]
+            # We have support for complex128
```
not complex64?

MichaelTiemannOSC replied:
I'd be happy to add complex64 to the test suite, but thought it better to do one case (complex128) at a time.  I didn't want to add something (`values.dtype.itemsize == 8`) that wasn't tested by the test suite.

jbrockmendel replied:
not a deal-breaker, but id prefer to do them both in this pass.  or at least leave a comment explaining why one is excluded

jbrockmendel commented on pandas/tests/arithmetic/test_numeric.py:
```diff
@@ -1116,45 +1116,75 @@ def test_ufunc_compat(self, holder, dtype):
+            exp_dtype = dtype
+        else:
+            assert result.dtype == "f8" and isinstance(result, box)
+            exp_dtype = np.float64
```
it looks like a bunch of these assertions are rendered redundant by the tm.assert_equal calls below them.  could we get away with just setting exp_dtype?  could even write a helper to de-duplicate these

MichaelTiemannOSC replied:
I think that would work.  I'll give it a try.

MichaelTiemannOSC committed:
Incorporate feedback

Updated code as per code review suggestions.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Update test_sparse.py

`test_fillna_no_op_returns_copy` now works for all cases in Pandas 3.0.0 without special casing.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

MichaelTiemannOSC commented:
The new 3.0.0 build rules make these changes happy.  Previous build rules across a variety of merge points in the 2.2.x created problems with inconsistent parameters unrelated to these changes.  Perhaps this is now ready to merge?

jbrockmendel commented on pandas/core/arrays/_mixins.py:
```diff
@@ -261,7 +261,13 @@ def shift(self, periods: int = 1, fill_value=None) -> Self:
+            # Note: when `self._ndarray.dtype.kind == "c"`, Numpy incorrectly complains
+            # that `must be real number, not ...` when in reality
+            # a complex argument is more likely what's expected
+            raise ValueError(exc.args) from exc
```
the comment here is helpful, thanks.  it suggests to me that we may want to only catch-and-re-raise in a subset of cases?  otherwise we'll be re-raising as ValueError more often than we really want?

MichaelTiemannOSC replied:
Previously @mroeschke questioned when ever this would be a TypeError in the first place...the above assignment was expected only ever raise a ValueError.  So, instead of trying to also handle TypeError in higher-level code, we translate this "impossible" case into the canonical form that Pandas expects.  The comment helps the user understand a completely unhelpful error message that comes from Python.  Previously I attempted to edit the error message to something more reasonable, but that was challenged.  At the end of the day the question is: how much steering do we want to do for this case vs. just letting the exception raise in the expected way and let users decipher what was wrong with their code in the first place.

MichaelTiemannOSC committed:
Merge branch 'main' into test_numpy_complex2

MichaelTiemannOSC committed:
Update algorithms.py

Allow all defined numpy complex dtypes to work as well as complex128.  Note that by default the test suite only tests the complex128 case, but users can add or alter that default by modifying the test suite source code to test other cases.

Signed-off-by: Michael Tiemann <72577720+MichaelTiemannOSC@users.noreply.github.com>

mroeschke commented:
Looks like this PR as gone stale so closing to clear the queue, if interested in contributing please merge in main and we can reopen

MichaelTiemannOSC commented:
Twice I've tried to get this merged, and twice it's gone stale.  I thought it was a good idea, and I'm willing to try again, but only if there's enough interest on the Pandas side to make complex128 a first-class test case.
